storage-proofs-porep/src/stacked/vanilla/challenges.rs

```diff
@@ -430,67 +361,42 @@ pub mod synthetic {
-            let first_prf_index = first_porep_challenge / prf_challenges;
-            let last_prf_index = last_porep_challenge / prf_challenges;
-
-            // If the first porep challenge's synthetic index bytes are not aligned to the first
-            // bytes generated by the prf, skip the leading bytes output by the first prf
-            // corresponding with challenges in the previous partition.
-            let skip_first_challenges = first_porep_challenge % prf_challenges;
-
-            let rand_bytes: Vec<u8> = if self.is_sha256() {
-                let sha256 =
-                    Hmac::<Sha256>::new_from_slice(b"filecoin.io|PoRep|1|Synthetic|1|Selection")
-                        .expect("should not fail")
-                        .chain_update(self.replica_id)
-                        .chain_update(rand);
-                (first_prf_index..=last_prf_index)
-                    .flat_map(|prf_index| {
-                        sha256
-                            .clone()
-                            .chain_update(&prf_index.to_le_bytes()[..4])
-                            .finalize()
-                            .into_bytes()
-                    })
-                    .collect()
-            } else {
-                let key = Blake2b::new()
-                    .hash_length(CHACHA20_KEY_SIZE)
-                    .key(b"filecoin.io|PoRep|1|Synthetic|1|Selection")
-                    .to_state()
-                    .update(&self.replica_id)
-                    .update(rand)
-                    .finalize();
-                let mut chacha20 = ChaCha20::new(key.as_bytes().into(), CHACHA20_NONCE.into());
-                let seek_pos = first_prf_index as u32 * CHACHA20_BLOCK_SIZE;
-                chacha20
-                    .try_seek(seek_pos)
-                    .expect("seek should not exceed keystream length");
-
-                (first_prf_index..=last_prf_index)
-                    .flat_map(|_| {
-                        let mut rand_bytes = [0u8; 64];
-                        chacha20.apply_keystream(&mut rand_bytes);
-                        rand_bytes.to_vec()
-                    })
-                    .collect()
-            };
 
-            rand_bytes
+            // The number of porep challenges generated per prf call.
+            let porep_challenges_per_block = 16;
+            // Indexes of chacha20 blocks used to generate partition `k`'s challenges.
+            let first_block = first_porep_challenge / porep_challenges_per_block;
+            let last_block = last_porep_challenge / porep_challenges_per_block;
+
+            // If the first porep challenge of the partition is not aligned to the first porep
+            // challenge generated by a chacha20 block, skip the leading challenges generated by the
+            // the first chacha20 block, i.e. skip the previous partition's last challenges.
+            let skip_first_challenges = first_porep_challenge % porep_challenges_per_block;
```

Comment:
We might be able to delete this as well if the chacha seeks allow you to seek in the middle of a block.

Reply:
If seeking to arbitrary byte position works for ChaCha here in Rust then you could seek to position: `first_porep_challenge*4` then `apply_keystream` to a buffer of size `4*num_partition_challenges`.